File: automator/src/main/java/com/stardust/automator/simple_action/FilterAction.java
package com.stardust.automator.simple_action;

import android.graphics.Rect;

import com.stardust.automator.UiObject;
import com.stardust.automator.filter.BoundsFilter;
import com.stardust.view.accessibility.AccessibilityNodeInfoHelper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by Stardust on 2017/1/27.
 */

public abstract class FilterAction extends SimpleAction {


    public interface Filter {

        List<UiObject> filter(UiObject root);
    }

    public static class TextFilter implements Filter {

        String mText;
        int mIndex;

        public TextFilter(String text, int index) {
            mText = text;
            mIndex = index;
        }

        @Override
        public List<UiObject> filter(UiObject root) {
            List<UiObject> list = root.findByText(mText);
            if (mIndex == -1)
                return list;
            if (mIndex >= list.size())
                return Collections.emptyList();
            return Collections.singletonList(list.get(mIndex));
        }

        @Override
        public String toString() {
            return "TextFilter{" +
                    "mText='" + mText + '\'' +
                    ", mIndex=" + mIndex +
                    '}';
        }
    }

    public static class BoundsFilter implements Filter {

        Rect mBoundsInScreen;

        public BoundsFilter(Rect boundsInScreen) {
            mBoundsInScreen = boundsInScreen;
        }

        @Override
        public List<UiObject> filter(UiObject root) {
            List<UiObject> list = new ArrayList<>();
            findAccessibilityNodeInfosByBounds(root, list);
            return list;
        }

        private void findAccessibilityNodeInfosByBounds(UiObject root, List<UiObject> list) {
            if (root == null)
                return;
            Rect rect = new Rect();
            root.getBoundsInScreen(rect);
            if (rect.equals(mBoundsInScreen)) {
                list.add(root);
            }
            int oldSize = list.size();
            for (int i = 0; i < root.getChildCount(); i++) {
                UiObject child = root.child(i);
                if (child == null)
                    continue;
                findAccessibilityNodeInfosByBounds(child, list);
            }
            if (oldSize == list.size() && rect.contains(mBoundsInScreen)) {
                list.add(root);
            }
        }

        @Override
        public String toString() {
            return "BoundsFilter{" +
                    "mBoundsInScreen=" + mBoundsInScreen +
                    '}';
        }
    }

    public static class EditableFilter implements Filter {

        private int mIndex;

        public EditableFilter(int index) {
            mIndex = index;
        }

        @Override
        public List<UiObject> filter(UiObject root) {
            List<UiObject> editableList = findEditable(root);
            if (mIndex == -1)
                return editableList;
            if (mIndex >= editableList.size())
                return Collections.emptyList();
            return Collections.singletonList(editableList.get(mIndex));
        }

        public static List<UiObject> findEditable(UiObject root) {
            if (root == null) {
                return Collections.emptyList();
            }
            if (root.isEditable()) {
                return Collections.singletonList(root);
            }
            List<UiObject> list = new LinkedList<>();
            for (int i = 0; i < root.getChildCount(); i++) {
                list.addAll(findEditable(root.child(i)));
            }
            return list;
        }

        @Override
        public String toString() {
            return "EditableFilter{" +
                    "mIndex=" + mIndex +
                    '}';
        }
    }

    public static class IdFilter implements Filter {

        private final String mId;

        public IdFilter(String id) {
            this.mId = id;
        }

        @Override
        public List<UiObject> filter(UiObject root) {
            return root.findByViewId(mId);
        }

        @Override
        public String toString() {
            return "IdFilter{" +
                    "mId='" + mId + '\'' +
                    '}';
        }
    }


    private Filter mFilter;

    public FilterAction(Filter filter) {
        mFilter = filter;
    }

    @Override
    public boolean perform(UiObject root) {
        if (root == null)
            return false;
        List<UiObject> list = mFilter.filter(root);
        return perform(list);
    }

    public abstract boolean perform(List<UiObject> nodes);

    public static class SimpleFilterAction extends FilterAction {

        private int mAction;

        public SimpleFilterAction(int action, Filter filter) {
            super(filter);
            mAction = action;
        }

        @Override
        public boolean perform(List<UiObject> nodes) {
            if (nodes == null || nodes.isEmpty())
                return false;
            boolean succeed = true;
            for (UiObject nodeInfo : nodes) {
                if (!nodeInfo.performAction(mAction)) {
                    succeed = false;
                }
            }
            return succeed;
        }
    }

    @Override
    public String toString() {
        return "FilterAction{" +
                "mFilter=" + mFilter +
                '}';
    }
}

File: automator/src/main/java/com/stardust/automator/simple_action/FilterAction.kt
package com.stardust.automator.simple_action

import android.graphics.Rect
import com.stardust.automator.UiObject
import java.util.*

/**
 * Created by Stardust on 2017/1/27.
 */

abstract class FilterAction(private val mFilter: Filter) : SimpleAction() {

    abstract fun perform(nodes: List<UiObject>): Boolean

    interface Filter {

        fun filter(root: UiObject): List<UiObject>
    }

    class TextFilter(private var mText: String, internal var mIndex: Int) : Filter {

        override fun filter(root: UiObject): List<UiObject> {
            val list = root.findByText(mText)
            if (mIndex == -1)
                return list
            return if (mIndex >= list.size) emptyList() else listOf(list[mIndex])
        }

        override fun toString(): String {
            return "TextFilter{" +
                    "mText='" + mText + '\''.toString() +
                    ", mIndex=" + mIndex +
                    '}'.toString()
        }
    }

    class BoundsFilter(private var mBoundsInScreen: Rect) : Filter {

        override fun filter(root: UiObject): List<UiObject> {
            val list = ArrayList<UiObject>()
            findAccessibilityNodeInfosByBounds(root, list)
            return list
        }

        private fun findAccessibilityNodeInfosByBounds(root: UiObject?, list: MutableList<UiObject>) {
            if (root == null)
                return
            val rect = Rect()
            root.getBoundsInScreen(rect)
            if (rect == mBoundsInScreen) {
                list.add(root)
            }
            val oldSize = list.size
            for (i in 0 until root.childCount) {
                val child = root.child(i) ?: continue
                findAccessibilityNodeInfosByBounds(child, list)
            }
            if (oldSize == list.size && rect.contains(mBoundsInScreen)) {
                list.add(root)
            }
        }

        override fun toString(): String {
            return "BoundsFilter{" +
                    "mBoundsInScreen=" + mBoundsInScreen +
                    '}'.toString()
        }
    }

    class EditableFilter(private val mIndex: Int) : Filter {

        override fun filter(root: UiObject): List<UiObject> {
            val editableList = findEditable(root)
            if (mIndex == -1)
                return editableList
            return if (mIndex >= editableList.size) emptyList() else listOf(editableList[mIndex])
        }

        override fun toString(): String {
            return "EditableFilter{" +
                    "mIndex=" + mIndex +
                    '}'.toString()
        }

        companion object {

            fun findEditable(root: UiObject?): List<UiObject> {
                if (root == null) {
                    return emptyList()
                }
                if (root.isEditable) {
                    return listOf(root)
                }
                val list = LinkedList<UiObject>()
                for (i in 0 until root.childCount) {
                    list.addAll(findEditable(root.child(i)))
                }
                return list
            }
        }
    }

    class IdFilter(private val mId: String) : Filter {

        override fun filter(root: UiObject): List<UiObject> {
            return root.findByViewId(mId)
        }

        override fun toString(): String {
            return "IdFilter{" +
                    "mId='" + mId + '\''.toString() +
                    '}'.toString()
        }
    }

    override fun perform(root: UiObject): Boolean {
        val list = mFilter.filter(root)
        return perform(list)
    }


    class SimpleFilterAction(private val mAction: Int, filter: Filter) : FilterAction(filter) {

        override fun perform(nodes: List<UiObject>): Boolean {
            if (nodes.isEmpty())
                return false
            var succeed = true
            for (nodeInfo in nodes) {
                if (!nodeInfo.performAction(mAction)) {
                    succeed = false
                }
            }
            return succeed
        }
    }

    override fun toString(): String {
        return "FilterAction{" +
                "mFilter=" + mFilter +
                '}'.toString()
    }
}
